docs/VerificationOptimization/VerificationOptimization.md

```diff
@@ -0,0 +1,625 @@
+
+<!-- %check-verify -->
+```dafny
+lemma LemmaMulEquality(x: int, y: int, z: int)
+  requires x == y
+  ensures x * z == y * z
+{}
+
+lemma LemmaMulEqualityAuto()
+  ensures forall x: int, y: int, z: int {:trigger x * z, y * z } :: x == y ==> x * z == y * z
+{
+  forall (x: int, y: int, z: int | x == y)
+    ensures x * z == y * z
+  {
+    LemmaMulEquality(x, y, z);
+  }
+}
+```
+
+The [`{:trigger}`](../DafnyRef/DafnyRef#sec-trigger) annotation tells it to only instantiate the formula in the postcondition when the solver sees two existing formulas, `x * z` and `y * z`. That is, if it sees two multiplications in which the right-hand sides are the same.
+
+Adding an explicit annotation is generally a last resort, however, because Dafny attempts to infer what triggers would be useful and typically comes up with an effective set. The need for manually-specified triggers often suggests that code could be refactored into a larger number of small, separate definitions, for which Dafny will then infer exactly the triggers you need.
+
+## Smaller, more isolated definitions
+
+As a general rule, designing a program from the start so that it can be broken into small pieces with clearly-defined interfaces is helpful for human understanding, flexibility, maintenance, and ease of verification. More specific instances of this include the following.
+
+* Break large functions into smaller functions, especially when they're ghost functions, and make at least some of them opaque. I often consider a ghost function of more than a handful of lines to be a code smell. There are exceptions where larger functions are necessary, or the most straightforward alternative, though, such when handling all of the many cases of a complex algebraic data type.
+
+* Prove lemmas about functions rather than having numerous or complex postconditions. The individual lemmas can then be invoked only when needed, and other facts that are true but irrelevant will not bog down the solver.
+
+* Break methods up into smaller pieces, as well. Because methods are always opaque, this will, in many cases, require adding a contract to each broken-out method. To make this more tractable, consider abstracting the concepts used in those contracts into separate functions, some of which may be opaque.
+
+# Summary
+
+When you're dealing with a program that exhibits high verification variability, or simply slow or unsuccessful verification, careful encapsulation and information hiding combined with hints to add information in the context of difficult constructs is frequently the best solution. Breaking one large definition with extensive contracts down into several, smaller definitions, each with simpler contracts, can be very valuable. Within each smaller definition, a few internal, locally-encapsulated hints to help the prover make progress in a more limited context can be very valuable.
```

Comment:
Slighty awk: ending two consecutive sentences with "can be very valuable" :)

Reply:
Another good catch.